text/0005-sapper-as-middleware.md

```diff
@@ -0,0 +1,165 @@
+- Start Date: 2020-08-31
+- RFC PR: 
+- Svelte Issue: 
+
+# Sapper as Middleware
+
+## Summary
+
+The Sapper build process should emit a mostly self-contained blob of javascript (and supporting files) that
+implements all server-side routes, SSR, and DOM code for the entire application. This javascript should
+export (at least) two functions:
+
+- A `middleware(req, res)` function that will serve a dynamic request.
+- A `assets(req, res)` function that will serve a static asset request.
+
+The supporting files *must* also include a manifest of files that can be uploaded to a CDN.
+
+This javascript *must not* start an Express server.
+
+## Motivation
+
+Working with Sapper in production for a large scale project has identified a number of
+challenges that arise from the unintended complexity of stale assumptions.
+
+By reducing these assumptions to the *absolute minumum*, we aim to:
+
+- Make it easier to deploy Sapper in efficient, scalable configurations
+- Make it easy to use Sapper applications in more scenarios than just a single standalone site
+- Make it easier to extend Sapper to suit specific project needs (without requiring modifications of Sapper itself)
+
+As of 0.28, the Sapper build process emits a completely standalone directory in `__sapper__/build`
+that starts an Express server listening on `$PORT` to serve the Sapper site.
+
+To do this Sapper makes many nuanced assumptions about the dependencies and runtime environment of
+applications it is building. Many of these assumptions are unnecessary sources of complexity during
+both developent and deployment.
+
+## Detailed design
+
+The general behavior of client-side bundling for Sapper should remain unchanged. Components are compiled for
+each route as SSR code and as DOM code. The behavior of the DOM (client-side) entry point is also unchanged.
+We expect the client-side bundle to be just as completely self-contained as it is today.
+
+This is in contrast with the server-side bundling operation. Instead of emitting a completely free-standing
```

Comment:
> it would instead emit a partially bundled .js file that the user should include in a script of their own

For the standard Express/Polka case that we support today, would the user do a second bundling operation or their own script could be run directly?

Reply:
It would be runnable directly without additional node dependencies, but if they were previously depending on sapper's bundling to make their server code easier to distribute they would need to switch to doing their own explicit bundle operation.